app/serializers/user.js

```diff
@@ -1,85 +1,89 @@
 // app/serializers/user.js
 import JSONAPISerializer from '@ember-data/serializer/json-api';
 
-/* 
-
-A serializer has two main roles in Ember Data. 
-First it is responsible for taking a response from an adapter and serializing 
-it into the normalized JSON format that Ember Data understands. 
-
-Secondly, it transforms snapshots of records into a payload the adapter 
-will send to the server when creating, updating or deleting a record. 
-
-*/
-
 export default class UserSerializer extends JSONAPISerializer {
+  // Define attribute mappings using the `attrs` property
+  attrs = {
+    firstName: 'first_name', // Map API `first_name` to model `firstName`
+    lastName: 'last_name', // Map API `last_name` to model `lastName`
+    description: 'description', // Maps directly to model attribute
+    tasks: { key: 'taskIds' }, // Map the `tasks` relationship to `taskIds`
+  };
+
+  /*
+   * This method normalizes the response from the API to ensure it fits the
+   * JSON API specification expected by Ember Data. It adjusts the payload
+   * for both creating a new user and retrieving existing users.
+   *
+   * When creating a user, the API only expects name and description fields.
+   * For retrieving users, the structure needs to conform to the JSON API
+   * format. Although adjusting the server would be ideal, handling
+   * API inconsistencies directly in the client is a more realistic approach.
+   */
   normalizeResponse(store, primaryModelClass, payload, id, requestType) {
-    let serializedData;
+    let users;
 
-    // Convert all IDs to strings to avoid type mismatches
-    const convertIdToString = (id) => id ? id.toString() : null;
+    // Create a function to map user data to the desired JSON API structure
+    const mappedUser = (user) => ({
+      type: 'user', // Specifies the type of resource
+      id: user.id.toString(), // Ensures the ID is a string
+      attributes: {
+        // Use the attrs mapping for dynamic access to API fields
+        // Square brackets compute the key names dynamically
+        // e.g., [this.attrs.firstName] translates to 'first_name'
+        // 'user[this.attrs.firstName]' retrieves the correct value from the user object
+        [this.attrs.firstName]: user[this.attrs.firstName],
+        [this.attrs.lastName]: user[this.attrs.lastName],
+        [this.attrs.description]: user[this.attrs.description],
+      },
+      relationships: {
+        tasks: {
+          // Use dynamic access for the task IDs
+          data: (user[this.attrs.tasks.key] || []).map((taskId) => ({
+            type: 'task',
+            id: taskId.toString(),
+          })),
+        },
+      },
+    });
```

Comment:
Do we really need it ? I thought that if we specify map (`attrs`) then we don't need mapping function :thinking:

Reply:
I did try to get it working just using the attrs but seemed to hit errors until I mapped the response to something more aligned with Ember Data, because my API response which is quite basic and not aligned with the Json that ember expects.

For example getting all users from my API provides this response:

{
    "users": [
        {
            "id": 1,
            "first_name": "Timothy",
            "last_name": "Boyd",
            "description": "Engineer",
            "taskIds": []
        },
        {
            "id": 2,
            "first_name": "Paul",
            "last_name": "Hughes",
            "description": "Farmer",
            "taskIds": []
        }
    ]
}

However, after looking at this again I agree I should probably be able to make it work without the mappedUser function and will take another look at it - I probably would still need to make sure my ids are strings (I think)

Reply:
Ive had another look, and i think this is required purely because my API response is not standard - perhaps I should change it as it would simplify a lot of things - I believe then I could get away with only using attrs:

So from this:
```
{
"users": [
    {
        "id": 1,
        "first_name": "Timothy",
        "last_name": "Boyd",
        "description": "Engineer",
        "taskIds": []
    },
    {
        "id": 2,
        "first_name": "Paul",
        "last_name": "Hughes",
        "description": "Farmer",
        "taskIds": []
    }
  ]
}
```


to something like this:

```
{
  "data": [
    {
      "type": "user",
      "id": "1",
      "attributes": {
        "first_name": "Timothy",
        "last_name": "Boyd",
        "description": "Engineer",
        "taskIds": []
      }
    },
    {
      "type": "user",
      "id": "2",
      "attributes": {
        "first_name": "Paul",
        "last_name": "Hughes",
        "description": "Farmer",
        "taskIds": []
      }
    }
  ]
}
```

Reply:
Ah, yes probably you are right, we need slightly different entire structure, which perhaps is not possible only with `attrs`